flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,255 @@
+
+/**
+ * A sink function that outputs to PubSub.
+ *
+ * @param <IN> type of PubSubSink messages to write
+ */
+public class PubSubSink<IN> extends RichSinkFunction<IN> implements CheckpointedFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSink.class);
+
+	private final Credentials credentials;
+	private final SerializationSchema<IN> serializationSchema;
+	private final String projectName;
+	private final String topicName;
+	private final String hostAndPortForEmulator;
+
+	private transient Publisher publisher;
+
+	private PubSubSink(
+		Credentials credentials,
+		SerializationSchema<IN> serializationSchema,
+		String projectName,
+		String topicName,
+		String hostAndPortForEmulator) {
+		this.credentials = credentials;
+		this.serializationSchema = serializationSchema;
+		this.projectName = projectName;
+		this.topicName = topicName;
+		this.hostAndPortForEmulator = hostAndPortForEmulator;
+	}
+
+	private transient ManagedChannel managedChannel = null;
+	private transient TransportChannel channel = null;
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+
+		if (hostAndPortForEmulator != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPortForEmulator)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel))
+					.setCredentialsProvider(NoCredentialsProvider.create());
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		shutdownPublisher();
+		shutdownTransportChannel();
+		shutdownManagedChannel();
+	}
+
+	private void shutdownPublisher() {
+		try {
+			if (publisher != null) {
+				publisher.shutdown();
+			}
+		} catch (Exception e) {
+			LOG.info("Shutting down Publisher failed.", e);
+		}
+	}
+
+	private void shutdownTransportChannel() {
+		if (channel == null) {
+			return;
+		}
+		try {
+			channel.close();
+		} catch (Exception e) {
+			LOG.info("Shutting down TransportChannel failed.", e);
+		}
+	}
+
+	private void shutdownManagedChannel() {
+		if (managedChannel == null) {
+			return;
+		}
+		try {
+			managedChannel.shutdownNow();
+			managedChannel.awaitTermination(1000L, TimeUnit.MILLISECONDS);
+		} catch (Exception e) {
+			LOG.info("Shutting down ManagedChannel failed.", e);
+		}
+	}
+
+	@Override
+	public void invoke(IN message, SinkFunction.Context context) throws Exception {
+		PubsubMessage pubsubMessage = PubsubMessage
+			.newBuilder()
+			.setData(ByteString.copyFrom(serializationSchema.serialize(message)))
+			.build();
+
+		publisher.publish(pubsubMessage).get();
```

Comment:
While the synchronous publish is simple, I am not sure if this is a performance compromise worth making. In checkpoint, a typical logic is following:

1. Flush & wait for all the handlers to be fired (a simple counter would do the job)
2. check if there has been any error reported by the handler ( a flag set by handler if some error occurs)
3. Proceed if there is no error, or fail if there is one.

Even if performance is not a critical concern in this PR, I still think we are losing too much here since such synchronous write causes one round-trip per message, which may become a usability problem.

Reply:
I agree, I've changed it so that we are now blocking / waiting in the snapshotState() function and if any exception has been thrown we rethrow it.